MoreLinq.Test/SliceTest.cs

```diff
@@ -122,7 +131,9 @@ public void TestSliceLongerThanSequence()
             var sequenceA = Enumerable.Range(1, count);
             var sequenceB = sequenceA.ToList();
             var resultA = sequenceA.Slice(count / 2, count);
-            var resultB = sequenceB.Slice(count / 2, count);
+            // .NET 8 introduced `Slice` on `List<T>` so use `AsEnumerable` to force the extension
+            // method to be invoked.
+            var resultB = sequenceB.AsEnumerable().Slice(count / 2, count);
```

Comment:
Would recommend consistency in the various tests here. Either all should use .AsEnumerable() or all should disable ide0057. Up to you which you think is cleaner.

Reply:
Agree. Lots of code added over time by different contributors needs a consistency refresh.

Reply:
> Agree. Lots of code added over time by different contributors needs a consistency refresh.

I'm talking about the code you added here. Please pick one method or the other for all three changes in this file.

Reply:
The others are marked as not applicable (N/A). Smells like a bug in the analyser? There is no overload of [`Slice` in MoreLINQ](https://morelinq.github.io/4.0/ref/api/html/M_MoreLinq_MoreEnumerable_Slice__1.htm) that allow use of the range operator.

Reply:
> The others are marked as not applicable (N/A). Smells like a bug in the analyser? There is no overload of [`Slice` in MoreLINQ](https://morelinq.github.io/4.0/ref/api/html/M_MoreLinq_MoreEnumerable_Slice__1.htm) that allow use of the range operator.

They are correct, and not a bug in the analyzer. You have to switch to net8 to see why they're correct. The issue is the same as described in the comment here, `List<>` added a `.Slice()` method in net8, which is accessible via the range operator. There are four tests in this file that all do `var x = y.ToList(); x.Slice(..);`. In each case, either IDE0057 needs to be suppressed in net8+, or resultB needs to use `sequenceB.AsEnumerable().Slice(...)` to switch to the MoreLinq method instead of the `List<>` method. 

Either is correct; but the same method should be used for all four tests in this file. I would recommend `.AsEnumerable()`, which is what you see in four lines in #1039.

Reply:
I was switched to .NET 8 target in the IDE, but I guess I misinterpreted something when working in a rush to get to a successful build state yesterday. Good to have two eyeballs (and fixed with 6abc4715ea1dd19d71736663c9b2963432bef24a), but going back and reading the tests in more detail today, I think there's lots wrong still. For one thing, `Enumerable.Range` in .NET 8 no longer returns a pure sequence, but [an object that is effectively a list](https://github.com/dotnet/runtime/blob/v8.0.0/src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs#L13) and therefore may be accidentally exercising the optimised path for many operators. That's the case here with `Slice` too. For example, the following test is supposed to test that the optimised list path behaves the same as the one where the source is a pure sequence, but both, `sequenceA` and `sequenceB`, end up being seen by `Slice` as lists effectively:

```c#
/// <summary>
/// Verify that a slice from 0 to sequence.Count() yields the source sequence.
/// NOTE: Since there are two different implementations of Slice() - one optimized that's
///       for lists and one that operates on any <c>IEnumerable{T}</c>, this test examines both.
/// </summary>
[Test]
public void TestSliceIdentity()
{
    const int count = 100;
    var sequenceA = Enumerable.Range(1, count);
    var sequenceB = sequenceA.ToList().AsEnumerable();

    var resultA = sequenceA.Slice(0, count);
    var resultB = sequenceB.Slice(0, count);

    Assert.That(resultA, Is.EqualTo(sequenceA));
    Assert.That(resultB, Is.EqualTo(sequenceB));
}
```

In fact, none of the tests except `TestSliceIsLazy` hit the following line:

https://github.com/morelinq/MoreLINQ/blob/97dbe9874072cdf149c21daa663c22a9de9779c6/MoreLinq/Slice.cs#L60

Reply:
All issues with `Slice` should be addressed via #1043.

Reply:
For `Slice` at least, we can test using the following:
```cs
    var sequenceA = Enumerable.Range(1, count).Select(x => x);
    var sequenceB = sequenceA.ToList();

    var resultA = MoreEnumerable.Slice(sequenceA, 0, count);
    var resultB = MoreEnumerable.Slice(sequenceB, 0, count);
```

The identity `Select` will use one of the `SelectIterator`s, which do implement some performance improvements, but do not implement `IList<>`. Using the `MoreEnumerable.Slice()` pattern will force calling the MoreLinq `.Slice()` method regardless of which methods may or may not exist on the sequence objects.

Reply:
Tracking this issue in #1044. It doesn't need to be solved in this PR since `Enumerable.Range(1, 10) is IList<int>` returns `false` up to .NET 7 and tests against the various targets will cover the other code path.